### ISSUE TITLE
Setting a random seed or passing a generator to Modules for deterministic sampling.

### ISSUE THREAD

**Issue by khatrishubham88** [pytorch/pytorch#65049]
## 🚀 Feature

I would like to obtain deterministic/reproducible weight initialization when working with say Con1d, while limiting the deterministic state to the weight [default] initialization part only and retaining the random state for the rest of the torch modules.

## Motivation

With this functionality the models can be in initialized in a reproducible fashion while still allowing randomness in shuffling and thereby the optimization path taken by the optimizer. This is particularly interesting when studying the properties of different optimizer and architectures.  

## Pitch

Following, non-functional, code expresses the idea I am trying to achieve

```
gen1 = torch.Generator()
gen1.manual_seed(123)
conv1 = nn.Conv1d(16, 33, 3, stride=2)
nn.init.kaiming_uniform_(conv.weight, a=0, mode='fan_in', nonlinearity='leaky_relu', generator=gen1)


gen2 = torch.Generator()
gen2.manual_seed(123)
conv2 = nn.Conv1d(16, 33, 3, stride=2)
nn.init.kaiming_uniform_(conv2.weight, a=0, mode='fan_in', nonlinearity='leaky_relu', generator=gen2)

assert conv2.weight == conv1.weight
```
Expected outcome: True

## Alternatives
Until now I have to use following work around which at times is tedious since I have to recreate the different initializers

```
gen1 = torch.Generator()
gen1.manual_seed(123)
conv1 = nn.Conv1d(16, 33, 3, stride=2)
bound = # some calculation to compute the bound for He Initialization
x = (-bound - bound) * torch.rand((33, 16, 3), generator=gen1) + bound
conv1.weight.data = x

gen2 = torch.Generator()
gen2.manual_seed(123)
conv2 = nn.Conv1d(16, 33, 3, stride=2)
bound = # some calculation to compute the bound for He Initialization
x = (-bound - bound) * torch.rand((33, 16, 3), generator=gen2) + bound
conv2.weight.data = x

assert torch.eq(conv2.weight.data, conv1.weight.data) 
```

**Comment by jbschlosser** [comment]
Hey @khatrishubham88, thanks for the detailed request! Seems like it'd be as simple as exposing the generator arg all the way up to the higher-level `torch.nn.init` functions.

**Comment by khatrishubham88** [comment]
@jbschlosser Indeed I am planning to do it myself. Just wanted to refine the requirement before I begin.

**Comment by khatrishubham88** [comment]
Anything specific I should look out for before I begin?

**Comment by khatrishubham88** [comment]
@jbschlosser Ideally this can also be exposed to all the nn module for initialization. Kind of takes away the need for manual initialization if the default initializer is used. For example a simple way to do stuff would be 
```
gen1 = torch.Generator()
gen1.manual_seed(123)
conv1 = nn.Conv1d(16, 33, 3, stride=2, generator=gen1)

gen2 = torch.Generator()
gen2.manual_seed(123)
conv2 = nn.Conv1d(16, 33, 3, stride=2, generator=gen2)

assert torch.eq(conv2.weight.data, conv1.weight.data) 
```

I would recommend bringing the functionality all the way to layer level.

**Comment by mruberry** [comment]
Packaging this initialization pattern in a function would probably be a reasonable work-around for now. That is, you could make your own wrapper constructor for the convolution module.

**Comment by khatrishubham88** [comment]
@mruberry Any specific reason to not bring this all the way upto core module level?

**Comment by mruberry** [comment]
> @mruberry Any specific reason to not bring this all the way upto core module level? I am not a big fan of hacks :P

I don't think it's a hack and we can also consider natively supporting this pattern while suggesting a workaround

**Comment by khatrishubham88** [comment]
well, maybe I should have used better term "workaround". 😅
Anyways, I just thought while I am at it might as well finish the job. But it is alright, I will start with init and see how it goes from there. 😉

**Comment by khatrishubham88** [comment]
@mruberry @jbschlosser  is there a dependencies list for ubuntu that are to be installed from apt store for building the package from source?

**Comment by mruberry** [comment]
> @mruberry @jbschlosser is there a dependencies list for ubuntu that are to be installed from apt store for building the package from source?

The PyTorch forums (https://discuss.pytorch.org/) are a good place for this kind of question (what you need may have already been answered there). I'm not aware of any dependencies except those described by the instructions for installing from source on GitHub.

**Comment by khatrishubham88** [comment]
@mruberry yeah while setting it up I realized the environment variable was not set right `CMAKE_BUILD_WITH_INSTALL_RPATH` was off by default.

Additionally it might be worth updating the doc at [Torch.uniform_](https://pytorch.org/docs/stable/generated/torch.Tensor.uniform_.html#torch.Tensor.uniform_), `uniform_` has generator functionality but it is not documented.

**Comment by jbschlosser** [comment]
> @jbschlosser Ideally this can also be exposed to all the nn module for initialization. Kind of takes away the need for manual initialization if the default initializer is used. For example a simple way to do stuff would be

Sounds nice for certain use cases, but I'm quite wary of making such widespread changes across torch.nn modules. If you'd like, it might be worth opening a new issue requesting this functionality for discussion.
